src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/SamClient.scala

```diff
@@ -0,0 +1,44 @@
+package org.broadinstitute.dsde.agora.server.dataaccess
+
+import java.util
+
+import akka.http.scaladsl.model.StatusCodes
+import org.broadinstitute.dsde.agora.server.AgoraConfig
+import org.broadinstitute.dsde.workbench.client.sam.api.UsersApi
+import org.broadinstitute.dsde.workbench.client.sam.model.UserStatusInfo
+import org.broadinstitute.dsde.workbench.client.sam.{ApiCallback, ApiClient, ApiException}
+
+import scala.concurrent.{Future, Promise}
+
+object SamClient {
+  private def samUserApi(accessToken: String) = {
+    val apiClient = new ApiClient()
+    apiClient.setAccessToken(accessToken)
+    apiClient.setBasePath(AgoraConfig.samUrl)
+    new UsersApi(apiClient)
+  }
+
+  def getUserEmail(token: String): Future[Option[String]] = {
+    val api = samUserApi(token)
+    val promise = Promise[Option[String]]()
```

Comment:
can you talk a bit about why we're using promises?

Reply:
https://docs.scala-lang.org/overviews/core/futures.html#promises

I don't want to use a Future because the sam client already is async. If I did
```
Future(api.getUserStatusInfoAsync(...))
```
Then I would be launching the call in the future, not waiting for it. What I really need is to do something when getUserStatusInfoAsync finishes. `Promise` provides a handy way to do that and convert the call into a `Future` which the rest of the code understands.

The sam client does have synchronous functions so I suppose I could also do
```
Future(api.getUserStatusInfo(...))
```
but then I need to think about what context the Future runs on. If it is on the main context then this using the `Future` does not really gain us anything since it still blocks threads that are not meant to be blocked. So I would need to make another context... and all of this would be silly anyway since the sam client is async under the hood and the sync call just has a wait in there.

Reply:
ah. i think my mistake here was assuming we wrote the client, and that the return type of `getUserStatusInfoAsync` was `Future`. neither of which is true (it's generated, and thus something insane)